tools/pyi/gen_pyi.py

```diff
@@ -1064,14 +1064,14 @@ def replace_special_case(hint: str) -> str:
             "new_tensor": [
                 f"def new_tensor(self, data: Any, {FACTORY_PARAMS}) -> Tensor: ..."
             ],
-            "__new__": ["def __new__(self, *args, **kwargs) -> Tensor: ..."],
+            "__new__": ["def __new__(cls, *args, **kwargs) -> Self: ..."],
```

Comment:
Hm, I tried with your example https://github.com/albanD/subclass_zoo/blob/main/uint4_tensor.py, and `Tensor.__new__(UInt4Tensor)` gives `UInt4Tensor([])` of type `<class 'uint4_tensor.UInt4Tensor'>`

Reply:
```python
from torch import Tensor

class MyTensor(torch.Tensor): ...

t = Tensor.__new__(MyTensor, [1,2,3])
assert type(t) == MyTensor  # ✅
```

I don't see the issue...

Reply:
I think it's reasonable to make sure we have accurate types for Tensor and Parameter only though, surely?!

Reply:
What I meant is this:
```python
from torch.nn import Parameter
from torch import Tensor

class MyTensor(Tensor): ...

t = Parameter(MyTensor([1,2,3]))
assert type(t) == MyTensor
assert type(t) != Parameter
```

Reply:
Right, here's what the patch changes, when using `mypy`/`pyright`:

```python
from torch.nn import Parameter
from torch import Tensor

class MyTensor(Tensor): ...

t = Parameter(MyTensor([1,2,3]))
reveal_type(t)  # before/after patch: mypy: Parameter/Parameter; pyright: Tensor/Parameter
assert type(t) == MyTensor
assert type(t) != Parameter

t2 = Parameter(Tensor([1,2,3]))
reveal_type(t2)  # before/after patch: mypy: Parameter/Parameter, pyright: Tensor/Parameter
assert type(t2) != Tensor
assert type(t2) == Parameter
```

The patch does not solve all problems, but it aligns the typing of `mypy` and `pyright` in the most important case. There is no behavioral change when using `mypy`.

I think solving the issue for `nn.Parameter` for custom Tensors is difficult with the current python type system, because the behavior violates the Liskov substitution principle and the type system does not cater to such cases.

Reply:
And again, to clarify: This is because `mypy` actually ignores the annotated return type of `__new__`, but `pyright` doesn't.

```python
class A:
    def __new__(cls) -> int: return 0
        
reveal_type(A())  # mypy: A, pyright: int
```

- [mypy-playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=e975c6d4cc64e6390f2fe6e6e067fd2f)
- [pyright-playground](https://pyright-play.net/?code=MYGwhgzhAECCBcAoaLoBMCmAzaB9XAdhgO74AUoEAlNALQB80AlgQC7zQBOGrArpwWgAGZKjGJuANwxgQuVgE8ADhjKwyVGtADE0ALbKFHWABpoShZyYBzABbtmbIA)
- https://github.com/python/mypy/issues/1020